### PR TITLE
load existing stats file when in AFL_AUTORESUME or -i - mode.

### PR DESCRIPTION
**Pull request by vj-27**
Hi, this is my first PR here :) 
It loads some of the variables in the stats file when using AFL_AUTORESUME or "-i -" . ( issue #274  )
It can load the following fields on status screen:
Overall Results:
- cycles done
- total paths
- uniq crashes
- uniq hangs
Findings in Depth:
- favored paths
Path Geometry:
- levels
- pending
- pend fav
- own finds
- imported

### CODE REVIEW

**Comment by vanhauser-thc** [comment]
at a first look this seems to be good!

if you import total execs done, then you will also need to load the run_time as otherwise the exec calculations will be wrong.

then - do not load any pending_* or *_favored - this is internally assessed.

**Comment by vj-27** [comment]
@vanhauser-thc I have a doubt.
if we load run_time, which variable to set? In the calculations, cur_time - start_time is used and the cur_time is set by calling gettimeofday(). or can i make a new variable prev_run_time in the state and use it for calculations (**prev_run_time +cur_time - start_time**) ?

**Comment by vanhauser-thc** [comment]
that would one option. easier would be just to subtract that value from start_time. the only disadvantage would be that start_time would be incorrect in the new fuzzer_stats but IMHO that is fine

**Comment by domenukk** [comment]
One other thing, how about an `AFL_RESUME_NOSTATS` or similar for people that want the old behavior?
Well thinking about it: they can just rm the stats file 🤔

**Comment by vj-27** [comment]
@domenukk sorry i didn't understand. :confused:

**Comment by vj-27** [comment]
> Why not handle the `pref_run_time` inline, like all the other vars?
> We just need to make sure the `afl->start_time` is set before calling this function

Can you check the latest commit?

File: src/afl-fuzz-stats.c [hunk 1]
```diff
@@ -89,6 +89,106 @@ void write_setup_file(afl_state_t *afl, u32 argc, char **argv) {
...
 
+/* load some of the existing stats file when resuming.*/
+void load_stats_file(afl_state_t *afl) {
+
+  FILE *f;
+  u8    buf[MAX_LINE];
+  u8 *  lptr;
+  u8    fn[PATH_MAX];
+  u32   lineno = 0;
+  afl->prev_run_time = 0;
```
Comment (vanhauser-thc) [inline src/afl-fuzz-stats.c]: the structure is null initialized, not needed

File: src/afl-fuzz-stats.c [hunk 2]
```diff
@@ -89,6 +89,106 @@ void write_setup_file(afl_state_t *afl, u32 argc, char **argv) {
...
+      strcpy(keystring, lstartptr);
+      lptr++;
+      char *nptr;
+      switch (lineno) {
+
+        case 3:
+          if (!strcmp(keystring, "run_time          ")) {
+
+            afl->prev_run_time = 1000 * strtoull(lptr, &nptr, 10);
+            afl->start_time -= afl->prev_run_time;
```
Comment (vanhauser-thc) [inline src/afl-fuzz-stats.c]: this makes the env var pointless, remove this line, then see other comment

File: src/afl-fuzz.c [hunk 3]
```diff
@@ -1700,6 +1703,8 @@ int main(int argc, char **argv_orig, char **envp) {
   // (void)nice(-20);  // does not improve the speed
   // real start time, we reset, so this works correctly with -V
   afl->start_time = get_cur_time();
+  if (afl->in_place_resume || afl->afl_env.afl_autoresume)
+    afl->start_time -= afl->prev_run_time;
```
Comment (vanhauser-thc) [inline src/afl-fuzz.c]: remove this line, this would make the var pointless

**Comment by vanhauser-thc** [comment]
what is missing is checking where start_time is used and plying the afl->prev_run_time var there, e.g.

```
afl-fuzz-stats.c:          "start_time        : %llu\n"
afl-fuzz-stats.c:          afl->start_time / 1000, cur_time / 1000,
```
would need to be
```
afl-fuzz-stats.c:          (afl->start_time - afl->prev_run_time var)  / 1000, cur_time / 1000,
```

but you have to check how start_time is used, this cannot be blindly replaced globally.

dominik is right that it needs afl->prev_run_time, because the value of start_time itself may not be modified.

**Comment by vj-27** [comment]
@vanhauser-thc thanks for the review
I have made the changes so that it affects only the stats screen and the stats file. 
And in the status screen when resuming, should the "run time " start from 0 or from prev_run_time ?

**Comment by vanhauser-thc** [comment]
there are a few more locations where prev_run_time has to be applied, but I take it from here.
Thank you for this PR!

**Pull request merged**